src/mesh/BoundingBox.cpp

```diff
@@ -0,0 +1,171 @@
+  }
+}
+
+BoundingBox::BoundingBox()
+{
+  _dimensions = 3;
+  for(int i = 0; i < _dimensions; ++i){
+    _bounds.push_back(std::numeric_limits<double>::max());
+    _bounds.push_back(std::numeric_limits<double>::lowest());
+  }
+}
+
+BoundingBox::BoundingBox(const BoundingBox &bb)
+{
+  _dimensions = bb._dimensions;
+  _bounds     = bb._bounds;
+}
+
+BoundingBox::~BoundingBox()
+{
+}
+
+std::ostream &operator<<(std::ostream &out, const BoundingBox &bb)
+{
+  for (int d = 0; d < bb._dimensions; ++d) {
+    out << "dim: " << d << " min: " << bb._bounds[2 * d] << ", max: " << bb._bounds[2 * d + 1] << "\n";
+  }
+  return out;
+}
+
+bool BoundingBox::operator==(const BoundingBox& otherBB) const
+{
+  if(_dimensions != otherBB._dimensions){
+    return false;
+  }
+  for(int i = 0; i < _dimensions; ++i){
+    if(std::abs(_bounds.at(i) - otherBB._bounds.at(i)) > 1e-6){
+      return false;
+    }
+  }
+  return true;
+}
+
+void BoundingBox::modifyForTest(int rank, std::string testName)
+{
+  if(testName == "com")
+  {
+    for (int i = 0; i < _dimensions; i++) {
+      _bounds[2*i] = rank*i;
+      _bounds[2*i + 1] = i + 1;
+    }
+  }
+  if(testName == "partition"){
+    for (int i = 0; i < _dimensions; i++) {
+      _bounds[2*i] = 3 - rank - 1;
+      _bounds[2*i + 1] = 3 - rank;
+    }
+  }
+}
+
+BoundingBox BoundingBox::createFromData(std::vector<double> bounds)
+{
+  BoundingBox box{bounds};
+  return box;
+}
+
+void BoundingBox::setSafetyFactor(double safetyFactor){
+  _safetyFactor = safetyFactor;
+}
+
+bool BoundingBox::isVertexInBB(const mesh::Vertex &vertex) const
+{
+  for (int d = 0; d < _dimensions; d++) {
+    if (vertex.getCoords()[d] < _bounds.at(2 * d) || vertex.getCoords()[d] > _bounds.at(2 * d + 1)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+double BoundingBox::getData(int dimension, int type) const
+{
+  return _bounds.at(2*dimension + (type - 1));
+}
+
+int BoundingBox::getDimension() const
+{
+  return _dimensions;
+}
+
+const double* BoundingBox::data() const
+{
+  return _bounds.data();
+}
+
+const std::vector<double> BoundingBox::dataVector() const
+{
+  return _bounds;
+}
+
+bool BoundingBox::mergeBoundingBoxes(const BoundingBox &otherBB)
+{
+  for (int d = 0; d < _dimensions; d++) {
+    _bounds[2 * d]     = std::min(_bounds[2 * d], otherBB._bounds[2 * d]);
+    _bounds[2 * d + 1] = std::max(_bounds[2 * d + 1], otherBB._bounds[2 * d + 1]);
+  }
+
+  // Enlarge BB
+  PRECICE_ASSERT(_safetyFactor >= 0.0);
+
+  double maxSideLength = 1e-6; // we need some minimum > 0 here
+
+  for (int d = 0; d < _dimensions; d++) {
+    if (_bounds.at(2 * d + 1) > _bounds.at(2 * d))
+      maxSideLength = std::max(maxSideLength, _bounds[2 * d + 1] - _bounds[2 * d]);
+  }
+  for (int d = 0; d < _dimensions; d++) {
+    _bounds.at(2 * d + 1) += _safetyFactor * maxSideLength;
+    _bounds.at(2 * d) -= _safetyFactor * maxSideLength;
+    PRECICE_DEBUG("Merged BoundingBox" << *this);
+  }
+  return true;
+}
```

Comment:
Can we split this in two functions. `expandTo(const BoundingBox &otherBB)` and `addSafetyMargin(double safetyFactor)`?
* `safetyFactor` could be passed as an argument here instead of the private member. Maybe that's easier to understand.
* I don't like the name `mergeBoundingBoxes` as it indicates a symmetric operation. But actually `otherBB` is not touched.